plexapi/library.py

```diff
@@ -549,16 +580,25 @@ def search(self, title=None, sort=None, maxresults=999999, libtype=None, **kwarg
             args['sort'] = self._cleanSearchSort(sort)
         if libtype is not None:
             args['type'] = utils.searchType(libtype)
-        # iterate over the results
-        results, subresults = [], '_init'
+
+        results = []
+        subresults = []
         args['X-Plex-Container-Start'] = 0
-        args['X-Plex-Container-Size'] = min(X_PLEX_CONTAINER_SIZE, maxresults)
-        while subresults and maxresults > len(results):
+        args['X-Plex-Container-Size'] = container_size
+        while True:
             key = '/library/sections/%s/all%s' % (self.key, utils.joinArgs(args))
             subresults = self.fetchItems(key)
-            results += subresults[:maxresults - len(results)]
+            results.extend(subresults)
+
+            # self.totalSize is not used as a condition in the while loop as
+            # this require a additional http request.
+            # self.totalSize is updated from .fetchItems
+            if self.totalSize <= len(results):
```

Comment:
IMO you should take `maxresults` into account here as well. Consider the use case where a library contains 5000 items and the user wants to retrieve the first 1000 in batches of 50 items. That would require 20 requests. Now you are making 100 requests and then throw away 4000 of the items.

Maybe something like this:
```python
maxresults = min(maxresults, self.totalSize)`
if maxresults <= len(results)
    break

args['X-Plex-Container-Start'] += args['X-Plex-Container-Size']
args['X-Plex-Container-Size'] = min(args['X-Plex-Container-Size'], maxresults - len(results))
```
This makes sure that you retrieve exactly the number of items requested. In case the user wants to retrieve all items you perform a few unnecessary calculations but those shouldn't be expensive compared to the (potentially unnecessary) HTTP requests.

And then you also don't need to slice the `results` list at the end.

Reply:
👍